keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+Kubernetes should define a set of standard metrics that allow a user who works with the Kubernetes resource model to easily answer via instrumentation several important queries:
+
+* What is the scale of my pending scheduling queue and what resources are blocking it
+* What is the current remaining schedulable capacity of all my nodes
+* Which components are using less than they request
+* Which components on a node are the most likely to be evicted
+* Do I have workloads that are requesting significantly more initialization resources than container resources in a way that is leading to underutilization
+* Which resources on my cluster are most contended
+* Is a set of pods requesting more resources than any node has to offer
+
+We should attempt to provide sufficient “out of the box” metrics for capacity planning for most cluster administrators, and present the resource model that Kubernetes uses for easy consumption. Integrations that extend the resource model with our supported pattern should also be visible in these metrics as the default resource model would interpret them.
+
+We should use consistent terminology, patterns, and descriptions of resource request, consumption, and availability across our API, metrics, component code, and documentation to ensure Kubernetes admins see a coherent view of the Kubernetes resource model.
+
+### Non-Goals
+
+* Full coverage all fields on pods in metrics
+* Container level representations of these metrics
+
+## Proposal
+
+### Report a new multi-dimension pod_resources metric
+
+A component with clear alignment to our resource consumption model should report a standard metric for pod consumption that takes into account lifecycle, resource type, units, and is consistent with the existing node_resources metric exposed by KSM.  This metric should be a pod level synthesis (resource decisions are pod level) of the same fundamental calculations the scheduler or kubelet make. It is recommended the scheduler expose this since it already has the relevant data, and custom schedulers may wish to expose similar metrics for their subset of usage. While the metrics are expected to cost significantly less than the set of metrics exposed by kube-state-metrics today for similar purposes (cardinality likely an order of magnitude lower), these metrics should be opt-in on the scheduler for those who have alternate or pre-existing resource metrics.
```

Comment:
To summarize and make sure I'm understanding: the goal/proposal here to move this small subset of instrumentation (pod_resources metrics) out of KSM and into the scheduler on an opt-in basis?

If that's correct, what would the implementation plan be? My reading here would be metrics that can be toggled with a CLI flag, but we should probably make this explicit.

Edit: this is now clear from reading further in the KEP, so it might be worth updating this section to preempt these questions.